File: LeetCode/0576/main.go
package main

import "fmt"

func findPaths(m int, n int, maxMove int, startRow int, startColumn int) int {
	MOD := 1000000007
	d := make([][][]int, m)
	for idx := range d {
		d[idx] = make([][]int, n)
		for id2 := range d[idx] {
			d[idx][id2] = make([]int, maxMove+1)
			for i:= 0; i < len(d[idx][id2]); i++ {
				d[idx][id2][i] = -1
			}
		}
	}
	var dfs func(x, y, rest int) int
	dfs = func(x, y, rest int) int {
		if x >= m || x < 0 || y >= n || y < 0 {
			return 1
		}
		if rest == 0 {
			return 0
		}
		if d[x][y][rest] > -1 {
			return d[x][y][rest]
		}
		for i := -1; i <= 1; i++ {
			for j := -1; j <= 1; j++ {
				if i+j == 1 || i+j == -1 {
					k := dfs(x+i, y+j, rest-1)
					if d[x][y][rest] == -1 { // 如果不单独判断的话，那么对于无法在rest步骤内走到边界外的(x,y)，其值总是-1
						d[x][y][rest] = k % MOD
					}else {
						d[x][y][rest] = (d[x][y][rest]+k)%MOD
					}
				}
			}
		}
		return d[x][y][rest]
	}
	return dfs(startRow, startColumn, maxMove)
}

func main() {
	fmt.Println(
		findPaths(45, 35, 47, 20, 31),
	)
}

File: LeetCode/0576/main.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<string.h>

typedef long long LL;
int p,q;
int MAX = 1<<31-1, MOD = 1000000007;
LL d[51][51][55];

LL dfs(int x, int y, int rest) {
    if (x < 0 || x >= p || y < 0 || y >= q) return 1;
    if (rest == 0) return 0;
    if (d[x][y][rest] != MAX) return d[x][y][rest];
    d[x][y][rest] = 0;
    d[x][y][rest] += 
        dfs(x+1,y,rest-1) + 
        dfs(x, y+1, rest-1) + 
        dfs(x-1,y,rest-1) + 
        dfs(x, y-1, rest-1);
    d[x][y][rest] %= MOD;
    return d[x][y][rest];
}

int findPaths(int m, int n, int maxMove, int startRow, int startColumn){
    p = m; q = n;
    int i, j, k;
    for (i = 0; i < m; i++)
        for (j = 0; j < n; j++) 
            for (k = 0; k < maxMove+1; k++)
                d[i][j][k] = MAX;
    return dfs(startRow, startColumn, maxMove);
}

int main(void) {
    printf("%d", findPaths(
        1,3,3,0,1
    ));
    return 0;
}